### PR TITLE
Insert headers in OwnCloudClient

### PR DESCRIPTION
**Pull request by abelgardep**
Fix https://github.com/owncloud/android/issues/2574
Needed for https://github.com/owncloud/android/pull/2913

### CODE REVIEW

File: owncloudComLibrary/src/main/java/com/owncloud/android/lib/common/http/methods/webdav/PropfindMethod.kt [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+/**
+ * Propfind calls wrapper
+ *
+ * @author David González Verdugo
+ */
+class PropfindMethod(
+    url: URL,
+    // request
+    val depth: Int,
+    private val mPropertiesToRequest: Array<Property.Name>
```
Comment (davigonz) [inline owncloudComLibrary/src/main/java/com/owncloud/android/lib/common/http/methods/webdav/PropfindMethod.kt:43]: I would remove the `m` prefix... that comes from the legacy java classes

File: owncloudComLibrary/src/main/java/com/owncloud/android/lib/common/http/methods/webdav/DavMethod.kt [hunk 2]
```diff
@@ -0,0 +1,164 @@
...
+import java.net.URL
+import java.util.concurrent.TimeUnit
+
+/**
+ * Wrapper to perform WebDAV (dav4android) calls
+ *
+ * @author David González Verdugo
+ */
+abstract class DavMethod protected constructor(url: URL) : HttpBaseMethod(url) {
+    protected var mDavResource: DavOCResource
```
Comment (davigonz) [inline owncloudComLibrary/src/main/java/com/owncloud/android/lib/common/http/methods/webdav/DavMethod.kt:47]: I would remove `m` prefix

File: owncloudComLibrary/src/main/java/com/owncloud/android/lib/common/http/methods/nonwebdav/PutMethod.kt [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+import okhttp3.RequestBody
+import java.io.IOException
+import java.net.URL
+
+/**
+ * OkHttp put calls wrapper
+ *
+ * @author David González Verdugo
+ */
+class PutMethod(url: URL, private val putRequestBody: RequestBody) : HttpMethod(url) {
```
Comment (davigonz) [inline owncloudComLibrary/src/main/java/com/owncloud/android/lib/common/http/methods/nonwebdav/PutMethod.kt:35]:
```suggestion
class PutMethod(
    url: URL,
    private val putRequestBody: RequestBody
) : HttpMethod(url) {
```

File: owncloudComLibrary/src/main/java/com/owncloud/android/lib/common/http/methods/webdav/DavMethod.kt [hunk 4]
```diff
@@ -0,0 +1,164 @@
...
+        mDavResource = DavOCResource(
+            okHttpClient,
+            httpUrl,
+            log
+        )
+    }
+
+    override fun setFollowRedirects(followRedirects: Boolean) {
+        super.setFollowRedirects(followRedirects)
+        mDavResource = DavOCResource(
```
Comment (davigonz) [inline owncloudComLibrary/src/main/java/com/owncloud/android/lib/common/http/methods/webdav/DavMethod.kt:120]: I do not particularly like this approach of creating new `DavOCResources`, I know that it's probably my code, I had a bad day I guess 😄  . 
What I mean is that we are updating the client via `super.setFollowRedirects(followRedirects)` but if you don't go there you won't see it and you won't know why we need a new DavOCResource here...

File: owncloudComLibrary/src/main/java/com/owncloud/android/lib/common/OwnCloudClient.java [hunk 5]
```diff
@@ -102,8 +97,17 @@ public int executeHttpMethod(HttpBaseMethod method) throws Exception {
...
-            setRequestId(method);
-
+            String requestId = RandomUtils.generateRandomUUID();
+
+            // Header to allow tracing requests in apache and ownCloud logs
+            Timber.d("Executing in request with id %s", requestId);
+            method.setRequestHeader(HttpConstants.OC_X_REQUEST_ID,requestId);
+            method.setRequestHeader(HttpConstants.USER_AGENT_HEADER, SingleSessionManager.getUserAgent());
+            method.setRequestHeader(HttpConstants.PARAM_SINGLE_COOKIE_HEADER, "true");
+            method.setRequestHeader(HttpConstants.ACCEPT_ENCODING_HEADER, HttpConstants.ACCEPT_ENCODING_IDENTITY);
```
Comment (davigonz) [inline owncloudComLibrary/src/main/java/com/owncloud/android/lib/common/OwnCloudClient.java:107]: What about creating a new `setRequestHeaders` method with variable args and write something like

```
method.setRequestHeaders(
    Pair(HttpConstants.OC_X_REQUEST_ID, requestId)
    Pair(HttpConstants. USER_AGENT_HEADER, SingleSessionManager.getUserAgent())
)
```

or even with a new RequestHeader object

```
method.setRequestHeaders(
    RequestHeader(HttpConstants.OC_X_REQUEST_ID, requestId)
    RequestHeader(HttpConstants. USER_AGENT_HEADER, SingleSessionManager.getUserAgent())
)
```

It's just a suggestion 🍺

File: owncloudComLibrary/src/main/java/com/owncloud/android/lib/common/OwnCloudClient.java [hunk 6]
```diff
@@ -102,8 +97,17 @@ public int executeHttpMethod(HttpBaseMethod method) throws Exception {
         int status;
 
         do {
-            setRequestId(method);
-
+            String requestId = RandomUtils.generateRandomUUID();
+
+            // Header to allow tracing requests in apache and ownCloud logs
+            Timber.d("Executing in request with id %s", requestId);
+            method.setRequestHeader(HttpConstants.OC_X_REQUEST_ID,requestId);
```
Comment (davigonz) [inline owncloudComLibrary/src/main/java/com/owncloud/android/lib/common/OwnCloudClient.java:104]: Space after `OC_X_REQUEST_ID`

File: owncloudComLibrary/build.gradle [hunk 7]
```diff
@@ -6,7 +6,7 @@ apply plugin: 'kotlin-allopen'
 dependencies {
     api 'com.squareup.okhttp3:okhttp:4.6.0'
     implementation "org.jetbrains.kotlin:kotlin-stdlib-jdk8:$kotlinVersion"
-    api 'com.gitlab.ownclouders:dav4android:oc_support_2.1'
+    api 'com.gitlab.ownclouders:dav4android:oc_support_2.1.2'
```
Comment (davigonz) [inline owncloudComLibrary/build.gradle:9]: ❤️

File: owncloudComLibrary/src/main/java/com/owncloud/android/lib/common/http/methods/webdav/PropfindMethod.kt [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+import java.util.ArrayList
+
+/**
+ * Propfind calls wrapper
+ *
+ * @author David González Verdugo
+ */
+class PropfindMethod(
+    url: URL,
+    // request
```
[the remark this answers is not in the document]
Comment (abelgardep) [reply owncloudComLibrary/src/main/java/com/owncloud/android/lib/common/http/methods/webdav/PropfindMethod.kt:41]: Removing it..

File: owncloudComLibrary/src/main/java/com/owncloud/android/lib/common/http/methods/HttpBaseMethod.kt [hunk 9]
```diff
@@ -0,0 +1,164 @@
...
+        return response.header(headerName)
+    }
+
+    // Body
+    fun getResponseBodyAsString(): String? {
+        if (responseBodyString == null && response.body != null) {
+            responseBodyString = response.body?.string()
+        }
+        return responseBodyString
+    }
```
[the remark this answers is not in the document]
Comment (abelgardep) [reply owncloudComLibrary/src/main/java/com/owncloud/android/lib/common/http/methods/HttpBaseMethod.kt:108]: Sure!

File: owncloudComLibrary/src/main/java/com/owncloud/android/lib/common/http/methods/HttpBaseMethod.kt [hunk 10]
```diff
@@ -0,0 +1,164 @@
...
+import java.net.MalformedURLException
+import java.net.URL
+import java.util.concurrent.TimeUnit
+
+abstract class HttpBaseMethod constructor(url: URL) {
+    var okHttpClient: OkHttpClient
+    var httpUrl: HttpUrl = url.toHttpUrlOrNull() ?: throw MalformedURLException()
+    var request: Request
+    var requestBody: RequestBody? = null
+    lateinit var response: Response
```
[the remark this answers is not in the document]
Comment (abelgardep) [reply owncloudComLibrary/src/main/java/com/owncloud/android/lib/common/http/methods/HttpBaseMethod.kt:22]: Could be an option, for sure. I will include it.

File: owncloudComLibrary/src/main/java/com/owncloud/android/lib/common/OwnCloudClient.java (hunk 5 above)
Comment (abelgardep) [reply owncloudComLibrary/src/main/java/com/owncloud/android/lib/common/OwnCloudClient.java]: Could be an option, but i didn't want to perform a big refactor. We could include that when we convert OwnCloudClient to kotlin.

File: owncloudComLibrary/src/main/java/com/owncloud/android/lib/common/http/methods/webdav/PutMethod.kt [hunk 11]
```diff
@@ -0,0 +1,50 @@
...
+ *
+ * @author David González Verdugo
+ */
+class PutMethod(
+    url: URL
+) : DavMethod(url) {
+    @Throws(IOException::class, HttpException::class)
+    public override fun onExecute(): Int {
+        mDavResource.put(
+            requestBody!!,
```
[the remark this answers is not in the document]
Comment (abelgardep) [reply owncloudComLibrary/src/main/java/com/owncloud/android/lib/common/http/methods/webdav/PutMethod.kt:42]: For sure, good catch 🔝

**Pull request merged**